### ISSUE TITLE
Support parsing and analyzing heap dumps

### ISSUE THREAD

**Issue by derekparker** [go-delve/delve#2655]
Go can produce heap dumps on demand. I think Delve should be able to read and analyze them and produce useful information for the user.

https://pkg.go.dev/runtime/debug#WriteHeapDump
https://github.com/golang/go/wiki/heapdump15-through-heapdump17

**Comment by hyangah** [comment]
cc @randall77

**Comment by jayantxie** [comment]
I have tried viewcore, but considering Delve as the de facto standard Go debugger, I plan to integrate heap analysis functionality into the Delve tool to provide a more stable and user-friendly memory analysis tool. cc @aarzilli 

Is there any update on the support for this proposal? Actually, I am currently working on the development of this.

**Comment by aarzilli** [comment]
Viewing heap memory is a fundamentally different thing than debugging. That is not to say that we couldn't do it (or shouldn't do it) but we would have to integrate the code of viewcore.

The problem with viewcore is that it deeply depends on go runtime internals and often doesn't get updated when they change. Moving it into delve wouldn't change this problem at all.

**Comment by jayantxie** [comment]
Is it possible to leverage the active community of Delve to maintain this functionality in the long term, thereby addressing the stability issues that have arisen due to the historically inactive maintenance of the viewcore tool? 

Or is the current plan of our community to rely on the APIs which are planned to be provided by viewcore for heap analysis integration? https://github.com/golang/go/issues/57447

I would prefer Delve to support viewing heap memory, as it would make problem troubleshooting more convenient compared to using viewcore separately.

**Comment by aarzilli** [comment]
Using dwarf for heap analysis is likely the wrong approach because it won't be able to analyze things kept alive through unsafe.Pointer like the GC does. Before 1.23 it won't be able to analyze closures correctly either. There's probably other edge cases it doesn't cover that I haven't thought about.
A naive implementation using dwarf is also probably excruciatingly slow.
This thing needs to be done the way viewcore does it, which means it needs to be similar to the runtime's GC. And the problem is keeping it in sync with the runtime through the years.

**Comment by jayantxie** [comment]
@aarzilli Is there a better way to determine the names of each object besides using dwarf? I noticed that the implementation of viewcore is also based on dwarf types, but it may still miss some information. Currently, I use the method of traversing pointers every 8 bytes to access and track pointers of unsafe.Pointer type. Go also uses unsafe.Pointer in sync.Map, and after some special handling, I can still identify the types of referenced sub-objects. Based on my experiments, using delve's locals+global vars as an initial value for recursive lookup seems to be a feasible approach. I am still working on optimizing performance, but I believe other methods would also face these issues.

**Comment by jayantxie** [comment]
Currently, my approach is to start from the locals+global vars and for each accessed sub-object, if it exists within the stack range or heap, I find the heap base+size. For fields with types that can be determined, I traverse them using dwarf type. Otherwise, I traverse them using `isPtrFromHeap`. By doing this, I can both maximize the clarity of object types and include unknown types of memory in the statistics.

**Comment by aarzilli** [comment]
> Is there a better way to determine the names of each object besides using dwarf? I noticed that the implementation of viewcore is also based on dwarf types

For that there are other ways but using dwarf could be reasonable. I wouldn't use it for anything else, however, since you can't use it for everything.

**Comment by jayantxie** [comment]
@aarzilli I agree with what you said. Delve may have its own considerations, and integrating object references into the tool could add significant complexity for the sake of maintainability. However, it is true that Golang lacks this feature, which would be very helpful for identifying many memory-related issues. Is there any good way to release this functionality without having to maintain it in my own forked repository? (Due to stability concerns with viewcore and the issue of missing some important information, I am not inclined to maintain this feature based on viewcore.)

**Comment by aarzilli** [comment]
> Is there any good way to release this functionality without having to maintain it in my own forked repository?

What would you be forking?

**Comment by jayantxie** [comment]
I forked the repository of Delve to develop a new subcommand: `objref ${filename}`. This is also the input for the `go tool pprof` command mentioned above.

**Comment by aarzilli** [comment]
I think the best way to do it would be to initially write your own main program outside of delve, possibly using delve's packages to parse core files, that does this. You wouldn't have to fork anything. The benefit of having this integrated into delve is small compared to having it at all.

If it proves to be maintainable and you release it with a license compatible with delve we can then think about making it into a delve subcommand.

**Comment by jayantxie** [comment]
@aarzilli I found it's not so easy to use delve as a library. There are many necessary APIs that are unexposed. Can I raise a PR to expose them?

**Comment by aarzilli** [comment]
Sure

**Comment by jayantxie** [comment]
https://github.com/cloudwego/goref

**Comment by kiashok** [comment]
> https://github.com/cloudwego/goref I have open-sourced a reference analysis tool based on delve that I developed recently to the cloudwego org. The main implementation involves using dwarf to scan types and referenced memory. For cases where dwarf scanning is not feasible, use gcmask scanning to cover them.

@jayantxie this is a great tool! Thanks for sharing!
